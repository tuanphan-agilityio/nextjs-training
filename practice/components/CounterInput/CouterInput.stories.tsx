import { action } from '@storybook/addon-actions';

import { Story, Meta } from '@storybook/react';

import CounterInput, { CounterInputProps } from '.';

export default {
  title: 'CounterInput',
  component: CounterInput,
} as Meta;

const Template: Story<CounterInputProps> = (args) => (
  <CounterInput onValueChange={action('Value changed')} {...args} />
);

export const Default = Template.bind({});
Default.args = {};

export const WithCustomInitialValue = Template.bind({});
WithCustomInitialValue.args = {
  initialValue: 5,
  minValue: 0,
  maxValue: 10,
};
