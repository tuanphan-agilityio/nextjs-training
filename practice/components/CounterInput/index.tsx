import { ChangeEvent, FC, memo, useCallback, useState } from 'react';
import clsx from 'clsx';

import Button from '@/components/Button';
import Input from '@/components/Input';

interface CounterInputProps {
  initialValue?: number;
  onValueChange?: (newValue: number) => void;
  minValue?: number;
  maxValue?: number;
}

const buttonStyles = clsx(
  'bg-none text-secondary font-primary-bold hover:bg-quinary sm:h-7',
);

const CounterInput: FC<CounterInputProps> = ({
  initialValue = 0,
  onValueChange = () => {},
  maxValue = Number.MAX_VALUE,
  minValue = 0,
}) => {
  const [count, setCount] = useState(initialValue);

  const handleIncrement = useCallback(() => {
    const newCount = count + 1;
    if (newCount <= maxValue) {
      setCount(newCount);
      onValueChange(newCount);
    }
  }, [count, onValueChange, maxValue]);

  const handleDecrement = useCallback(() => {
    const newCount = count - 1;
    if (newCount >= minValue) {
      setCount(newCount);
      onValueChange(newCount);
    }
  }, [count, onValueChange, minValue]);

  const handleInputChange = useCallback(
    (event: ChangeEvent<HTMLInputElement>) => {
      const inputValue = event.target.value;

      if (inputValue === '') {
        setCount(minValue);
        onValueChange(minValue);
      } else {
        const newValue = parseInt(inputValue, 10);
        if (!isNaN(newValue) && newValue >= minValue && newValue <= maxValue) {
          setCount(newValue);
          onValueChange(newValue);
        }
      }
    },
    [onValueChange, minValue, maxValue],
  );

  const isMinValue = count === minValue;
  const isMaxValue = count === maxValue;

  return (
    <div className='flex items-center p-1 w-[142px] font-secondary-bold border border-secondary bg-primary sm:h-8'>
      <Button
        onClick={handleDecrement}
        className={clsx(buttonStyles, isMinValue && 'hover:cursor-no-drop')}
      >
        -
      </Button>
      <Input
        align='center'
        value={count}
        onChange={handleInputChange}
        className='sm:h-7'
        aria-label='counter number'
      />
      <Button
        onClick={handleIncrement}
        className={clsx(buttonStyles, isMaxValue && 'hover:cursor-no-drop')}
      >
        +
      </Button>
    </div>
  );
};

export type { CounterInputProps };

export default memo(CounterInput);
