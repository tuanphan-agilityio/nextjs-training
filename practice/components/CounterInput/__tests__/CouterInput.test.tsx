import { render, screen, fireEvent } from '@testing-library/react';

import CounterInput, { CounterInputProps } from '..';

describe('CounterInput Component', () => {
  const mockProps: Required<CounterInputProps> = {
    initialValue: 3,
    onValueChange: jest.fn(),
    minValue: 0,
    maxValue: 10,
  };

  it('renders the initial value correctly', () => {
    render(<CounterInput {...mockProps} />);
    const input = screen.getByRole('textbox');

    expect(input).toHaveValue(mockProps.initialValue.toString());
  });

  it('increments the value when the "+" button is clicked', () => {
    render(<CounterInput {...mockProps} />);
    const incrementButton = screen.getByText('+');
    fireEvent.click(incrementButton);

    expect(mockProps.onValueChange).toHaveBeenCalledWith(
      mockProps.initialValue + 1,
    );
  });

  it('decrements the value when the "-" button is clicked', () => {
    render(<CounterInput {...mockProps} />);
    const decrementButton = screen.getByText('-');
    fireEvent.click(decrementButton);

    expect(mockProps.onValueChange).toHaveBeenCalledWith(
      mockProps.initialValue - 1,
    );
  });

  it('changes the value when a valid input is entered', () => {
    render(<CounterInput {...mockProps} />);
    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { value: '5' } });

    expect(mockProps.onValueChange).toHaveBeenCalledWith(5);
  });

  it('matches the snapshot', () => {
    const { container } = render(<CounterInput {...mockProps} />);
    expect(container).toMatchSnapshot();
  });

  it('should not decrement below min value', () => {
    render(<CounterInput {...mockProps} />);

    const decrementButton = screen.getByText('-');
    const inputValue = screen.getByDisplayValue(
      mockProps.initialValue.toString(),
    );

    for (let i = 0; i < 10; i++) {
      fireEvent.click(decrementButton);
    }

    expect(inputValue).toHaveValue(mockProps.minValue.toString());
  });

  it('should not increment above max value', () => {
    render(<CounterInput {...mockProps} />);

    const incrementButton = screen.getByText('+');
    const inputValue = screen.getByDisplayValue(
      mockProps.initialValue.toString(),
    );

    for (let i = 0; i < 10; i++) {
      fireEvent.click(incrementButton);
    }

    expect(inputValue).toHaveValue(mockProps.maxValue.toString());
  });
});
