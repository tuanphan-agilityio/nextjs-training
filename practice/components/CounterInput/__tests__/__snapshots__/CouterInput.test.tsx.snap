// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CounterInput Component matches the snapshot 1`] = `
<div>
  <div
    class="flex items-center p-1 w-[142px] font-secondary-bold border border-secondary bg-primary"
  >
    <button
      class="inline-flex justify-center items-center w-[180px] h-12 bg-primary font-primary-regular text-md hover:opacity-80 bg-gradient-to-b from-primary-gradient-start to-primary-gradient-end text-primary bg-none text-secondary font-primary-bold hover:bg-quinary"
    >
      -
    </button>
    
    <div
      class="flex px-4 py-2 w-full bg-primary"
    >
      <input
        class="block w-full border-none focus:outline-none text-center"
        value="3"
      />
    </div>
    <button
      class="inline-flex justify-center items-center w-[180px] h-12 bg-primary font-primary-regular text-md hover:opacity-80 bg-gradient-to-b from-primary-gradient-start to-primary-gradient-end text-primary bg-none text-secondary font-primary-bold hover:bg-quinary"
    >
      +
    </button>
  </div>
</div>
`;
